### PR TITLE
Rndbot talents reset bug

### PR DESCRIPTION
**Pull request by avirar**
Added check for bot persistence when bot->resetTalents is called in PlayerbotFactory::Randomize.

Enabled detection of 0/0/0 talents during PlayerbotFactory::Refresh. When a bot above the persistence level is found with unallocated talents (0/0/0), it will automatically select the optimal specialization by evaluating the cumulative gear score across all available specs and choosing the one with the highest alignment to its current equipment.

The 0/0/0 talent auto-fix is far from perfect as the scores aren't normalised across specs. It works well enough and ensures their gear/spec combination is viable.

### CODE REVIEW

**Comment by avirar** [comment]
#544

**Comment by hermensbas** [comment]
Isnt possible to reuse existing functions around autogear and init?

**Comment by avirar** [comment]
Since the bug only appeards with equipmentPersistence I have made keeping the bots existing gear a priority. This isn't generating new gear for the bot, it's selecting a spec based on the gear they already have, and only if it detects 0/0/0 after a call to initTalentTrees. The bug that causes the wipe issue is resolved by putting resetTalents inside the if statement for equipmentPersistance, but that does not resolve the issue with bots that previously had their talents wiped since all of the calls to initTalentTree during regular bot activity require equipmentPersistence to be disabled to function.

**Comment by hermensbas** [comment]
I cant help to wonder whether you are trying to solve an side-effect. The this states only occurs when bots reached maxed level not when leveling, also some are not able to reproduce.

Shouldnt we fix the core issue itself, why are bots at max level randomly losing talents? Or i am missing something?

File: src/factory/PlayerbotFactory.cpp [hunk 1]
```diff
@@ -207,7 +207,10 @@ void PlayerbotFactory::Randomize(bool incremental)
     Prepare();
     LOG_DEBUG("playerbots", "Resetting player...");
     PerformanceMonitorOperation* pmo = sPerformanceMonitor->start(PERF_MON_RNDBOT, "PlayerbotFactory_Reset");
-    bot->resetTalents(true);
+    if (!sPlayerbotAIConfig->equipmentPersistence || bot->GetLevel() < sPlayerbotAIConfig->equipmentPersistenceLevel)
+    {
+        bot->resetTalents(true);
+    }
```
Comment (liyunfan1223) [inline src/factory/PlayerbotFactory.cpp:213]: Modifying only these lines should solve this issue. Other modifications should be unnecessary.

**Pull request closed without merging**